gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java

```diff
@@ -268,4 +445,53 @@ private void testElementTriggers(BatchingSettings settings) throws Exception {
     assertThat(result.get()).isEqualTo(16);
     assertThat(anotherResult.isDone()).isTrue();
   }
+
+  private class ElementAddThread extends Thread {
+    private volatile boolean isCancelled;
+    private final AtomicInteger counter = new AtomicInteger();
+    private final Batcher<Integer, Integer> batcher;
+
+    ElementAddThread(Batcher<Integer, Integer> batcher) {
+      this.batcher = batcher;
+    }
+
+    @Override
+    public void run() {
+      while (!isCancelled) {
+        try {
+          batcher.add(counter.incrementAndGet());
+        } catch (Throwable ex) {
+          cancel();
+          throw new AssertionError("Caught exception while adding element");
+        }
+      }
+    }
+
+    void cancel() {
+      isCancelled = true;
+    }
+  }
+
+  private class BatchTestCallable extends UnaryCallable<LabeledIntList, List<Integer>> {
+
+    private volatile boolean isDuplicate;
+    final Set<Integer> treeSet = new HashSet<>();
+
+    @Override
+    public ApiFuture<List<Integer>> futureCall(LabeledIntList request, ApiCallContext context) {
```

Comment:
This will be called from multiple threads so the element collector needs to be threadsafe

Reply:
I am not sure what would be the preferred way,  I could use `Collections.synchronizedSet()` as well `synchronized` block with lock for `set.contains(..)` and `set.addAll(...)`.

I have updated this method with the latter approach. Please let me know if this looks fine to you.

Reply:
I would just use a concurrentmap & putIfAbsent